### ISSUE TITLE
fetch-ocsp-response succeeds, but the response is "Responder Error: trylater (3)"

### ISSUE THREAD

**Issue by tatsuhiro-t** [h2o/h2o#449]
I just encountered this today that fetch-ocsp-response script succeeds, but the response from OCSP responder was "Response Error: trylater (3)", and it looks like an error.
The resulting OCSP data, which are written in stdout, was just 5 bytes.

I don't know the result is always 5 bytes, if this error is occurred.
Observing successful case, the good result is around 1400 bytes, but this may be depends on certificate or something (I didn't fully read ocsp response, sorry!)

Perhaps, we should check "Responder Error:" message in openssl command output, and bail out?

Reading relevant OpenSSL source code:
```
$ grep "Responder Error" *.c apps/
ocsp.c:        BIO_printf(out, "Responder Error: %s (%d)\n",
```
I found this
```c
    i = OCSP_response_status(resp);

    if (i != OCSP_RESPONSE_STATUS_SUCCESSFUL) {
        BIO_printf(out, "Responder Error: %s (%d)\n",
                   OCSP_response_status_str(i), i);
        if (ignore_err)
            goto redo_accept;
        ret = 0;
        goto end;
    }
```
So you see "ret = 0"  on error.   But looking around the code, it seems that 1 is assigned to ret on error case.  So this could be OpenSSL bug.  I have not checked latest OpenSSL repo yet.

Generally speaking, not everyone upgrades OpenSSL to the latest version, but OS vendor will backport security fixes.  Not sure this potential bug is "security issue."
So most people may continue to use unpatched OpenSSL for some time, which means we have to do something if we'd like to handle this issue gracefully.

What do you think about this issue and how to handle it?

**Comment by tatsuhiro-t** [comment]
Alright, I checked latest OpenSSL code, and it still has this issue: https://github.com/openssl/openssl/blob/3da9505dc02b0594633c73a11343f54bb5dbf536/apps/ocsp.c#L660

**Comment by tatsuhiro-t** [comment]
I sent bug report mail to openssl.  Wait and see, what they respond with it.

Related to openssl ocsp command, I found that openssl ocsp command bundled with openssl <= 1.0.1 returns exit code 0 even if ocsp response verification failed.
1.0.1: https://github.com/openssl/openssl/blob/2507c8cfb3664cfd9bf94e597e83afb4646451de/apps/ocsp.c#L818
1.0.2: https://github.com/openssl/openssl/blob/b012b497eaac76da42ef6741d50c3c17ec8ed732/apps/ocsp.c#L855

This bug was fixed in 1.0.2 series.
So perhaps, we should also check that verification command does not emit "Response Verify Failure"?

**Comment by tatsuhiro-t** [comment]
In the first post, I wrote "5 bytes" for error message.  This is the der encoded valid response, but status code is 3 (trylater).

**Comment by kazuho** [comment]
@tatsuhiro-t 

Thank you for noticing the issue, tracking down the cause, and for communicating with the vendors.

> Perhaps, we should check "Responder Error:" message in openssl command output, and bail out?
> 
> Reading relevant OpenSSL source code:
> ```
> $ grep "Responder Error" *.c apps/ 
> ocsp.c: BIO_printf(out, "Responder Error: %s (%d)\n",
> ```

This looks bad for two reasons, one that it exits with zero, and the other is that the error message is emitted to stdout (instead of stderr).
Fortunately a valid OCSP response will never start from `R`, so yes, I agree that doing such check on stdout can be a workaround.

> So perhaps, we should also check that verification command does not emit "Response Verify Failure"?

Agreed.

Can you create a PR fixing the issues?  I would highly appreciate it.

**Comment by tatsuhiro-t** [comment]
Sure.  See #454.
I'm not really a perl programmer, so be sure that my code really works for you.
BTW, we nghttp2 project translated fetch-ocsp-response script in python: https://github.com/tatsuhiro-t/nghttp2/blob/647e30619fd689747c71dfa046d6f529c97ba097/script/fetch-ocsp-response
And it had the same fix.

**Comment by kazuho** [comment]
Thank you for the PR.  Will merge soon.

> BTW, we nghttp2 project translated fetch-ocsp-response script into python

I did not know that.  My expectation was that it'd be easier for you to fix since you were facing the issue in the Perl version.

**Comment by tatsuhiro-t** [comment]
Thank you.
Actually, I faced this issue with our python version.  And I also checked
original perl version and found that it also has this issue.
2015/08/25 8:55 "Kazuho Oku" <notifications@github.com>:

**Comment by kazuho** [comment]
Fixed in 838d855.
